### PR TITLE
Partial SwiftSyntax support for experimental `@abi` attribute

### PR DESCRIPTION
**Pull request by beccadax**
Matches (or will match) https://github.com/swiftlang/swift/pull/76878.

This is *mostly* a subset of the work in #2882 which adds support for `@abi` on most of the declarations I intend to permit, but not on nominal types. It's intended to unblock progress on the feature while I work through how to refactor the syntax tree for nominal types without making overly disruptive changes.

I've updated a few things here to respond to review comments. In particular, a review comment asking about `#if` handling prompted me to add a test for that; I found that I didn't like how it was being diagnosed, so I added a facility that can parse `#if` syntax into unexpected nodes on its contents.

### CODE REVIEW

**Comment by beccadax** [comment]
@swift-ci please test

File: Sources/SwiftParser/Directives.swift [hunk 1]
```diff
@@ -207,6 +219,38 @@ extension Parser {
...
+    // Consume `#if` and condition.
+    let poundIfToken = self.consumeAnyToken()
+    let conditionTokens = self.consumeRemainingTokenOnLine()
+    let unexpectedBefore = RawUnexpectedNodesSyntax(combining: poundIfToken, conditionTokens, arena: self.arena)!
+
+    // Construct contextual unexpected node provider that will be used to consume the rest.
+    enclosingUnexpected.addUnexpected(before: unexpectedBefore, afterThrough: .nestingLevel(of: self))
+
+    // Parse the node we actually want.
+    return parseResult(&self, enclosingUnexpected)
```
[the remark this answers is not in the document]
Comment (beccadax) [reply Sources/SwiftParser/Directives.swift:252]: Ha—I actually did something like that first, but decided it required too many hacks. (Particularly when I added in support for decl headers, which could only be smuggled through the outer `IfConfigDeclSyntax` by wrapping them in the corresponding decl...)

Fine, though, I'll work it out.

**Comment by beccadax** [comment]
@swift-ci please test

File: Sources/SwiftParser/Directives.swift (hunk 1 above)
[the remark this answers is not in the document]
Comment (beccadax) [reply Sources/SwiftParser/Directives.swift]: Changed the implementation to something that "un-parses" the `RawIfConfigDeclSyntax` nodes after they've been parsed.

**Review by ahoppen — APPROVED** [review]
Thank you, Becca. I like this, all the non-trivial stuff is nicely localized and thus easy to reason about.

**Comment by nkcsgexi** [comment]
@swift-ci please test

**Comment by nkcsgexi** [comment]
`testSwiftValidationTestsuite` seems to be failing on macOS. @beccadax could you take a look?

**Comment by ahoppen** [comment]
That failure is unrelated. I’m investigating it.

**Comment by ahoppen** [comment]
Disabled `testSwiftValidationTestsuite` while I investigate further: https://github.com/swiftlang/swift-syntax/pull/2909

@swift-ci Please test macOS

**Comment by nkcsgexi** [comment]
Thank you, Alex!

**Comment by beccadax** [comment]
@ahoppen The validation test failure may be something I saw when I was working on this PR: `validation-test/compiler_crashers_fixed/26170-swift-parser-skipsingle.swift` was causing a stack overflow because the (invalid) brace pairs were too deeply nested. I've worked around it by applying this locally:

```diff
--- a/Tests/SwiftParserTest/ParserTests.swift
+++ b/Tests/SwiftParserTest/ParserTests.swift
@@ -134,7 +134,13 @@ class ParserTests: ParserTestCase {
     runParserTests(
       name: "Swift validation tests",
       path: testDir,
-      checkDiagnostics: false
+      checkDiagnostics: false,
+      shouldExclude: { url in
+        [
+          // FIXME: This test is overflowing the stack
+          "26170-swift-parser-skipsingle.swift"
+        ].contains(url.lastPathComponent)
+      }
     )
   }
 }
```

**Comment by nkcsgexi** [comment]
🎉

**Pull request merged**